### DISCUSSION TITLE
Zombie Processes with uvicorn multiple workers shutdown

### DISCUSSION THREAD

**Discussion by polizogo67** [Kludex/uvicorn#2512] [Potential Issue]
Hi,

I decides that it would be better to use multiple workers with uvicorn for the scalability of my app.
I observed that every time I shut down the server there is a zombie process hanging on the system.

Ubuntu 20.04.6 LTS
Python 3.11.9
uvicorn 0.32.0
fastapi 0.112.0

```python
# test_app.py

from fastapi import FastAPI
app = FastAPI()

@app.get('/ping')
async def ping_pong():
    return {'msg': "pong"}
```

```python
# test_server.py
import uvicorn
import multiprocessing

if __name__ == '__main__':
    multiprocessing.set_start_method('fork')  # Set start method for multiprocessing
    multiprocessing.freeze_support()
    uvicorn.run("test_app:app", host="localhost", port=8081, reload=False, log_level="info", workers=4)
```

After a few restarts I have a bunch of zombie processes that only disappear by restarting the system.
It seems like a serious issue for production environments.
![image](https://github.com/user-attachments/assets/56416a82-5dfb-4bf1-8491-af6ab2fca8eb)

**Comment by zffocussss** [comment]
You can use the uvicorn worker for your multiple processes usage.It is so strange to use the multipleprocessing lib directly.

**Reply by polizogo67** [reply]
Ι am not using multiprocessing, I am just setting the freeze support because if not there is WARNING/ERROR happening which says to turn it on.

Anyway I have the exact same behavior without multiprocessing.

```python
import uvicorn
if __name__ == '__main__':
    uvicorn.run("test_app:app", host="localhost", port=8081, reload=False, log_level="info", workers=4)
```

**Comment by polizogo67** [comment]
Any Updates on that ?
